[develop] Split EnsembleStat vx tasks into GenEnsProd and EnsembleStat

gsketefian opened the pull request:
## DESCRIPTION OF CHANGES: 
<!-- One or more paragraphs describing the problem, solution, and required changes. -->
The latest versions of MET/METplus have introduced a new tool named `GenEnsProd` that performs many of the ensemble-related calculations that `EnsembleStat` used to do.  Although `EnsembleStat` can still perform these calculations, that will change in future versions of MET/METplus.  Thus, this PR creates new vx tasks that call `GenEnsProd` to perform the necessary calculations, removing them from the `EnsembleStat` tasks.  Details:
* Add new METplus config files for `GenEnsProd`.
* Modify `exregional_run_met_genensprod_or_ensemblestat.sh` to enable it to call `GenEnsProd` (as well as `EnsembleStat`).
* Modify the `EnsembleStat` config files so that `EnsembleStat` no longer calculates the ensemble statistics that are now the purview of `GenEnsProd`.
* Modify the ex-scripts for the ensemble mean and probabilistic tasks so that they look for forecast input from the output of the `GenEnsProd` tasks, not the output of the `EnsembleStat` tasks.
* Modify all the METplus config files so that they are as similar to each other as possible (for possible consolidation later on).
* For clarity and correctness, rename the workflow configuration files `verify.yaml` and `verify_ensgrid.yaml` to `verify_det.yaml` and `verify_ens.yaml`, respectively.
* Add new yaml configuration file named `verify_pre.yaml` that contains (meta)tasks that are prerequisites for both deterministic and ensemble vx (these tasks were previously included in `verify_det.yaml`). 
* Modify the default ensemble vx workflow configuration file `verify_ens.yaml` to include the new `GenEnsProd` tasks.
* Modify `verify_det.yaml` and `verify_ens.yaml` so that vx task dependencies can be formulated more precisely, e.g. have dependencies on member-specific tasks instead of whole metatasks.  This includes splitting metatasks into two or more other ones, moving tasks from one metatask to another, etc.
* Modify the WE2E configuration files for the four vx tests (`MET_verification`, `MET_verification_only_vx`, `MET_ensemble_verification`, and `MET_ensemble_verification_only_vx`) to reflect the changes in the default vx workflow configuration files `verify_pre.yaml`, `verify_det.yaml`, and `verify_ens.yaml`.
* Introduce the new experiment variable `VX_FIELDS` to enable the user to specify the fields or groups of fields on which to perform verification.
* Introduce the new experiment variable `VX_APCP_ACCUMS_HH` to enable the user to specify the accumulation hours for which to perform verification on accumulated precip (APCP).
* Modify `setup.py` to remove any [meta]tasks for which fields are not specified in `VX_FIELDS`.
* Modify `setup.py` so that for configuration parameters that are arrays/lists, each value of the list is checked to ensure it is valid (if that variable has a corresponding `valid_vals_` array in `valid_param_vals.yaml`).
* Make modifications to vx ex-scripts as well as `config_defaults.sh` to ensure that the vx tests succeed in NCO mode.
* Make sure that the experiment variable `ENS_TIME_LAG_HRS` is set and accessed correctly for both deterministic and ensemble configurations.  For a deterministic configuration, this is a 1-element array containing the time lag (if any, in units of hours), while for an ensemble configuration it contains `NUM_ENS_MEMBERS` elements containing the time lag for each element.
* Add a new WE2E test (`MET_ensemble_verification_only_vx_time_lag`) that performs deterministic and ensemble verification on a 2-member ensemble with one time-lagged and one non-time-lagged member.
* Add the existing WE2E test `MET_ensemble_verification_only_vx` and the new test `MET_ensemble_verification_only_vx_time_lag` to the list of comprehensive WE2E tests.
* Include hour 0 in the verification of MRMS and NDAS fields.
* Fix bugs in `exregional_get_obs_[ccpa|mrms|ndas].sh` ex-scripts.

### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [x] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

I ran the following vx WE2E tests on Hera with intel, both in community mode and in NCO mode:
```
MET_verification
MET_verification_only_vx
MET_ensemble_verification
MET_ensemble_verification_only_vx
MET_ensemble_verification_only_vx_time_lag
```
I also ran the following `fundamental` suite of tests:
```
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8_plot
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16
```
All tests were successful.

## DOCUMENTATION:
<!-- If this PR is contributing new capabilities that need to be documented, please also include updates to the RST files (docs/UsersGuide/source) as supporting material. -->
Documentation is needed but will be added once the whole set of vx PRs has been merged (see Issue #630).

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

gsketefian committed:
Cleanup of EnsembleStat configuration files.

gsketefian committed:
Clean up comments in METplus config files.

gsketefian committed:
Standardize METplus conf files as much as possible.

gsketefian committed:
Fix up comments in METplus conf files.

gsketefian committed:
Simplificatons to EnsembleStat conf files.

gsketefian committed:
Add GenEnsProd tasks to verification workflow.

gsketefian committed:
Remove lines in EnsembleStat conf files that are for actions that are now done in GenEnsProd (to avoid duplication).

gsketefian committed:
Change input to ensemble mean and probabilistic tasks to come from the output of GenEnsProd (instead of EnsembleStat, which no longer generates the necessary outputs).

gsketefian committed:
Rename metatasks/tasks for clarity; create new metatasks/tasks to make desired dependencies possible; remove unneeded variables; change vx WE2E test config files to be compatible with aforementioned changes.

gsketefian committed:
Add config variable VX_APCP_ACCUMS_HH to make the accumulation hours for which to perform vx user-specifiable.

gsketefian committed:
Add the experiment configuration parameter VX_FIELDS that specifies which fields or group of fields on which to perform verification.

gsketefian committed:
Rename and add vx metatasks for clarity.

gsketefian committed:
Bug fix in metatask name.

gsketefian committed:
Move CCPA/APCP metatask to before MRMS and NDAS once since the former field(s) are usually come first in vx.

gsketefian committed:
Remove rocoto verification metatasks from workflow when there are no fields for those metatasks.

gsketefian committed:
Clean up vx [meta]task dependencies.

gsketefian committed:
Add new experiment flags RUN_VX_DET and RUN_VX_ENS that specify whether or not to run deterministic and ensemble verification, respectively.

gsketefian committed:
Bug fix for GEPES_MRMS metatask.

gsketefian committed:
Bug fixes to ensemble verification task dependencies.

gsketefian committed:
Bug fixes to ensemble task dependencies.

gsketefian committed:
Modifications to vx WE2E tests that don't use staged forecasts to get them to work properly with latest changes.

gsketefian committed:
Rename vx workflow yaml configuration file names for clarity and accuracy.

gsketefian committed:
For consistency with the way other variables are passed into the METplus config files (and in order for the final config files to have the correct values), change the way the variable ENSMEM_INDX is passed into METplus config files from an environment variable to a jinja variable.

gsketefian committed:
Merge branch 'develop' into feature/split_gepes

gsketefian committed:
Changes to make vx tests work in NCO mode:

1) Replace "mem${ensmem_indx}" with new variable "${ensmem_name}".
2) Remove DOT_ENSMEM_OR_NULL since ensmem_name can now be used instead.
3) Replace SLASH_ENSMEM_SUBDIR_OR_NULL with slash_ensmem_subdir_or_null.
4) Remove unneeded code from exregional_run_met_pb2nc_obs.sh.

gsketefian committed:
Change variable name for clarity.

gsketefian committed:
Add code to allow each element of a list variable to be checked for valid values; specify valid sets of values for VX_FIELDS and VX_APCP_ACCUMS_HRS.

christinaholtNOAA commented on parm/wflow/verify_ens.yaml:
```diff
@@ -23,56 +22,18 @@ default_task_verify_ens: &default_task_verify_ens
   walltime: 01:00:00
 
-metatask_EnsembleStat:
+metatask_GEPES_CCPA:
```
I'd suggest spelling this out to something like `metatask_GenEnsProd_CCPA`. Nobody knows what GEPES is and this particular piece of information isn't really referenced a ton to require a bunch of keystrokes. It could help with usability.

gsketefian replied:
@christinaholtNOAA `GEPES` is short for something like `GenEnsProd_EnsembleStat`, which I thought was too long to include as a metatask name.  But I'm ok with it if you are!

gsketefian replied:
@christinaholtNOAA Changed.

christinaholtNOAA commented on ush/config_defaults.yaml:
```diff
@@ -2427,12 +2427,40 @@ verification:
+  #
+  RUN_VX_DET: False
+  RUN_VX_ENS: False
+  #
```
In general, I think we should avoid including these types of workflow switches. Perhaps it makes sense to break out the PcpCombine* and Pb2nc* task so that it's easier to specify running those. Then all the user has to do is specify which groups they'd like to run instead of using these flags to turn off a group of items.

For example, since we have these flags, the user needs to set two different things in the config file to even turn on ensemble verification...the task group must be added AND the RUN_VX_ENS flag must be set to True. If these flags were removed, the only thing they'd need to do is include the verify_ens.yaml file in task groups.

Since I likely misunderstood the assignment when grouping the PcpCombine* and Pb2nc* into the verify_det.yaml group, and those may also be needed when running ensemble vx, maybe they belong in their own taskgroup like `prepare_vx_obs` or something similar. Then the user chooses to turn these tasks on/off by including/not including them in the task group instead of having this second switch that will give un-intended results. 

Mainly what I'm concerned with is a user saying "I turned on this switch, why is it not working?" and them also needing to know that they must flip another switch.

gsketefian replied:
@christinaholtNOAA Yes, somehow I got it stuck in my head that I shouldn't further split the `verify_[det|ens].yaml` files, but that's the best solution to avoiding these flags.  It was also bothering me that the user has to change two things in the config file to get vx to work right.  Thanks for the suggestion, I'll change it.

gsketefian replied:
@christinaholtNOAA Now the tasks that are prerequisites for both deterministic and ensemble vx are in a yaml file of their own (`verify_pre.yaml`), and the `RUN_VX_[DET|ENS]` switches are removed from `config_defaults.yaml`.

christinaholtNOAA commented on ush/setup.py:
```diff
@@ -540,6 +540,95 @@ def remove_tag(tasks, tag):
+                        from workflow since no fields belonging to observation type "{obstype}"
+                        are specified for verification."""
+                    ))
+                    rocoto_config['tasks'].pop(metatask, None)
```
The logic below here would not be needed if the switches are removed as suggested above.

gsketefian replied:
@christinaholtNOAA Yep!  Will remove as part of next set of changes.

gsketefian replied:
@christinaholtNOAA Removed.

christinaholtNOAA commented on tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR.yaml:
```diff
@@ -35,3 +35,4 @@ task_get_extrn_lbcs:
   USE_USER_STAGED_EXTRN_FILES: true
 verification:
   VX_FCST_MODEL_NAME: FV3_GFS_v15p2_CONUS_25km
+  RUN_VX_DET: True
```
Here's a really nice example of what I mean in the longer comment. To get this to work out you had to add 2 separate pieces of information so that setup.py doesn't remove all the verify_det tasks.

gsketefian replied:
@christinaholtNOAA I agree.

gsketefian replied:
@christinaholtNOAA  The `RUN_VX_[DET|ENS]` switches have been removed.

christinaholtNOAA reviewed (commented):
@gsketefian I left a couple of comments below. I also wanted to draw your attention to Issue #633 that has a goal of adopting a more standard naming convention. This PR might be a good place to start given it touches a lot of the MET ex-scripts.

gsketefian committed:
Change argument list to function set_vx_fhr_list() for clarity and to allow separate max number of missing files for obs vs. forecast.

gsketefian committed:
Minor spacing edits.

gsketefian committed:
Split out vx preprocessing tasks that are needed by both deterministic and ensemble vx tasks into a new workflow task group file (verify_pre.yaml); add new vx preprocessing metatask (check_post_output) that checks that all post-processed forecast output files exist; fix up vx task dependencies.

gsketefian committed:
Remove experiment variables RUN_VX_DET and RUN_VX_ENS since their functionality is replaced by inclusion/exclusion of workflow taskgroup yaml configuration files.

gsketefian committed:
If no vx fields are specified, remove the metatask "check_post_output_all_mems" from the workflow.

gsketefian committed:
Clean up check of consistency between DO_ENSEMBLE flag and inclusion of ensemble vx tasks in the workflow (by making sure that the names of the ensemble vx tasks are read in from the appropriate yaml file instead of being hard-coded).

gsketefian committed:
Fix the way the time-lag array ENS_TIME_LAG_HRS is set (for both ensemble and deterministic forecasts) and the way the time lag is calculated for a given ensemble member (or the deterministic forecast).

gsketefian commented:
@christinaholtNOAA I think I addressed all of your concerns except the renaming of the vx tasks (Issue #633). 
 In the interests of getting this PR in so others who are working on further mods to vx can start their work and to not complicate this PR any further, I'd like to address that in a separate PR (also I'd like to consult @michelleharrold, @JeffBeck-NOAA, @willmayfield, and @mkavulich before I rename to make sure we agree).  Hope that is ok with you.  Thanks.

gsketefian committed:
Add missing files.

JeffBeck-NOAA commented:
> @christinaholtNOAA I think I addressed all of your concerns except the renaming of the vx tasks (Issue #633). In the interests of getting this PR in so others who are working on further mods to vx can start their work and to not complicate this PR any further, I'd like to address that in a separate PR (also I'd like to consult @michelleharrold, @JeffBeck-NOAA, @willmayfield, and @mkavulich before I rename to make sure we agree). Hope that is ok with you. Thanks.

@gsketefian, thanks for all your work on this PR!  I agree that we can tackle the renaming of the vx tasks in a separate PR after we consult with others.  I'd personally be fine with just removing the "run" portion of the ex-script and j-job names for all vx tasks.

gsketefian committed:
Modifications to get verification to work on time-lagged staged forecast files.

gsketefian committed:
Add WE2E test for time-lagging.

gsketefian committed:
Change the test configuration for the WE2E test "MET_ensemble_verification_only_vx_time_lag" to look for the staged forecast data from the default location for the machine; also, change the predefined grid name to the one on which this staged data is provided (this is necessary to form the correct path to the data).

gsketefian committed:
Add missing vx WE2E tests to the list of comprehensive tests.

gsketefian committed:
Merge branch 'develop' into feature/split_gepes

gsketefian committed:
Add forecast lead time (based on nominal cycle start time disregarding any time-lagging) to names of GenEnsProd and EnsembleStat output files.

gsketefian committed:
Bug fix: Change dependence of ensmean and ensprob tasks to be on the GenEnsProd task, not the EnsembleStat task.  Also, change order of metatasks so that the one for MRMS comes before the one for NDAS.

gsketefian committed:
For MRMS and NDAS fields, include hour 0 in the verification.

gsketefian committed:
Bug fix requested by Michelle in get_obs_... ex-scripts.

gsketefian committed:
Second bug fix requested by Michelle: add FHR back into get_obs_... tasks.

MichaelLueken reviewed (approved):
@gsketefian These changes look good to me!  I was also able to successfully run the new `MET_ensemble_verification_only_vx_time_lag` test, as well as the rest of the verification WE2E tests on Hera.

MichaelLueken commented:
@gsketefian I will go ahead and submit the automated Jenkins testing on this work now.  If @christinaholtNOAA requests additional changes, then the tests can be resubmitted at that time.

christinaholtNOAA reviewed (approved):
@gsketefian Thanks for addressing my concerns! This looks great.

MichaelLueken commented:
@gsketefian The automated Jenkins tests encountered failures:

Cheyenne GNU - the `nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km` test failed (detailed in issue #806).

Gaea - SRW won't build following the latest maintenance on the machine.  Will be turning off Gaea testing in #799 until the UFS-WM has merged the new Gaea modulefile changes.

Hera Intel - the `grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test failed (detailed in issue #805).

I saw that the there were two failures in the Hera GNU test - `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` and `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`.  These tests shouldn't be failing, so I have resubmitted them.  I will let you know if they continue to fail.

I will also run the Jenkins tests manually on Orion, since the machine requires git/2.28.0 to be loaded, but since git/2.28.0 isn't available on Hercules and Hercules and Orion are the same machine, the git/2.28.0 module load is constantly removed from the role-epic-ps account's .bashrc file.  I will let you know if there are any failures.

MichaelLueken commented:
@gsketefian Both the manual run of the Jenkins tests on Orion and the rerun of Hera Intel automated Jenkins tests have all successfully passed.  I will now move forward with merging this work to develop.

gsketefian commented:
@MichaelLueken Thanks for running the tests and merging!
